src/sagemaker_xgboost_container/algorithm_mode/callback.py

```diff
@@ -0,0 +1,205 @@
+            to 5 (that is, the 5 most recent checkpoint files are kept.)
+        start_iteration: indicates the round at which the current training
+            started. If xgb_model was loaded from a previous checkpoint, this
+            will be greater than 0 (that is, if the previous training ended
+            after round 19, start_iteration will be 20).
+        num_round: (optional) indicates the number of boosting rounds.
+        """
+        self.checkpoint_dir = checkpoint_dir
+        self.max_to_keep = max_to_keep
+        self.start_iteration = start_iteration
+        self.num_round = num_round
+
+        self.step = start_iteration
+        self.previous_checkpoints = [
+            os.path.join(self.checkpoint_dir, f)
+            for f in os.listdir(self.checkpoint_dir)]
+
+        self.thread = None
+        self.delete_queue = queue.Queue()
+
+        self.start()
+
+    def __call__(self, env: CallbackEnv) -> Callable[[CallbackEnv], None]:
+
+        if not os.path.exists(self.checkpoint_dir):
+            os.makedirs(self.checkpoint_dir)
+
+        return self.callback(env)
+
+    def fmt_path(self, i: int) -> str:
+
+        filename = "{}.{}".format(CHECKPOINT_FILENAME, _zero_pad(i))
+        checkpoint_path = os.path.join(self.checkpoint_dir, filename)
+
+        return checkpoint_path
+
+    def start(self) -> None:
+
+        def _delete_once(skip_locked_files=True):
+
+            for i in iter(self.delete_queue.get, self.SENTINEL):
+
+                path = self.fmt_path(i)
+                if (skip_locked_files
+                        and os.path.isfile(path + FILE_LOCK_SUFFIX)
+                        and not os.path.isfile(path + FILE_SAFE_SUFFIX)):
+                    self.delete_queue.put(i)
+                    continue
+                try:
+                    os.remove(path)
+                except Exception:
+                    logger.debug("Failed to delete %s", path)
+                finally:
+                    self.delete_queue.task_done()
+
+            self.delete_queue.task_done()
+
+        def _delete_twice():
+
+            _delete_once(skip_locked_files=True)
+            # Here, we've reached the end of training because we place sentinel in the
+            # queue at the end of training. We put another sentinel, go through everything
+            # in the queue once again, and try to remove it anyway where there is a lock on
+            # the file or not, because the training is done. On sagemaker, this should send
+            # a delete signal to the agent so that the upload can be canceled and removed
+            # from S3.
+            self.delete_queue.put(self.SENTINEL)
+            _delete_once(skip_locked_files=False)
+
+        self.thread = threading.Thread(target=_delete_twice, daemon=True)
+        self.thread.start()
+
+    def stop(self) -> None:
+
+        self.delete_queue.put(self.SENTINEL)
+        self.thread.join()
+
+    def callback(self, env: CallbackEnv) -> None:
```

Comment:
Can you describe what is being done here? Between the file suffixes and trying to keep up with the file names its a bit confusing.